File: Repository/ParticipantDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using Microsoft.EntityFrameworkCore.Metadata.Internal;
using Microsoft.EntityFrameworkCore.Query.Internal;
using mpp_proiect_csharp_DianaGliga11.Model;
using NLog.Fluent;

namespace mpp_proiect_csharp_DianaGliga11.Repository;

public class ParticipantDBRepository : I_ParticipantDBRepository
{
    
    private static readonly ILog log = LogManager.GetLogger(typeof(ParticipantDBRepository));
    private readonly IDictionary<string, string?> Props;
    public ParticipantDBRepository(IDictionary<string, string?> props) 
    {
        log.Info($"{nameof(ParticipantDBRepository)} constructed.");
        Props = props;
    }

    public void Add(Participant entity)
    {
        log.Info($"Adding Participant: {entity}");
        IDbConnection connection = DbConnectionUtils.GetConnection(Props);

        try
        {
            using var command = connection.CreateCommand();
            command.CommandText = @"INSERT INTO Participants (name, age) VALUES (@name, @age)";

            var nameParam = command.CreateParameter();
            nameParam.ParameterName = "@name";
            nameParam.Value = entity.Name;
            command.Parameters.Add(nameParam);

            var ageParam = command.CreateParameter();
            ageParam.ParameterName = "@age";
            ageParam.Value = entity.Age;
            command.Parameters.Add(ageParam);

            var result = command.ExecuteNonQuery();
        }
        catch (Exception ex)
        {
            log.Error("Error while adding new Participant ", ex);
            throw new EntityRepoException(ex);
        }

        Log.Info($"Added Participant: {entity}");
    }
    

    public void Remove(long id)
    {
        log.Info($"Removing Participant: {id}");
       IDbConnection connection = DbConnectionUtils.GetConnection(Props);
       try
       {
           using var command = connection.CreateCommand();
           command.CommandText = @"DELETE FROM Participants WHERE id = @id";
           var idParam = command.CreateParameter();
           idParam.ParameterName = "@id";
           idParam.Value =id;
           command.Parameters.Add(idParam);
           var result = command.ExecuteNonQuery();
       }
       catch (Exception ex)
       {
           log.Error("Error while removing Participant ", ex);
           throw new EntityRepoException(ex);
       }
       Log.Info($"Removed Participant: {id}");
    }

    public void Update(long id, Participant entity)
    {
        log.Info($"Updating Participant: {entity}");
        IDbConnection connection = DbConnectionUtils.GetConnection(Props);
        try
        {
            using var command = connection.CreateCommand();
            command.CommandText = "UPDATE Participants SET name = @name, age = @age WHERE id = @id";

            var nameParam = command.CreateParameter();
            nameParam.ParameterName = "@name";
            nameParam.Value = entity.Name;
            command.Parameters.Add(nameParam);
            
            var ageParam = command.CreateParameter();
            ageParam.ParameterName = "@age";
            ageParam.Value = entity.Age;
            command.Parameters.Add(ageParam);
            var result = command.ExecuteNonQuery();
        }
        catch (Exception ex)
        {
            log.Error("Error while updating Participant ", ex);
            throw new EntityRepoException(ex);
        }
        Log.Info($"Updated Participant: {entity}");
    }

    public Participant findById(long id)
    {
        log.Info($"Finding Participant: {id}");
        IDbConnection connection = DbConnectionUtils.GetConnection(Props);
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from \"Participants\" where \"Id\"=@id";
                var paramID = command.CreateParameter();
                paramID.ParameterName = "@id";
                paramID.Value = id;
                command.Parameters.Add(paramID);
                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        Participant participant = Extract(dataReader);
                        Log.Info("Exiting Finding Participant: {id}");
                        return participant;
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error("Participant was not found ", e);
            throw new EntityRepoException("Participant was not found");
        }

        return null;
    }

    public IEnumerable<Participant> getAll()
    {
        log.Info($"Getting All Participants");
        IDbConnection connection = DbConnectionUtils.GetConnection(Props);
        IList<Participant> participants = new List<Participant>();
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select * from \"Participants\"";
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Participant participant = Extract(dataReader);
                    participants.Add(participant);
                }
            }
        }
        log.Info("Exiting Get All Participants");
        return participants;
    }

    public Participant GetParticipantsByData(Participant participant)
    {
        log.Info($"Searching for Participant: {participant}");
        IDbConnection connection = DbConnectionUtils.GetConnection(Props);
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText =
                    "select * from \"Participants\" where \"Id\"=@id and name name=@name and age age=@age";
                var nameParam = command.CreateParameter();
                nameParam.ParameterName = "@name";
                nameParam.Value = participant.Name;
                command.Parameters.Add(nameParam);

                var ageParam = command.CreateParameter();
                ageParam.ParameterName = "@age";
                ageParam.Value = participant.Age;
                command.Parameters.Add(ageParam);
                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        Participant extracted = Extract(dataReader);
                        Log.Info("Exiting Finding Participant: {id}");
                        return extracted;
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error("Participant was not found ", e);
            throw new EntityRepoException("Participant was not found");
        }
        return null;
    }

    private Participant Extract(IDataReader dataReader)
    {
        var id = dataReader.GetInt64(0);
        var name = dataReader.GetString(1);
        var age = dataReader.GetInt32(2);
        var participant = new Participant(name, age);
        participant.Id = id;
        return participant;
    }
}

File: Repository/src/main/java/example/repo/ParticipantDBRepository.java
package example.repo;

import Utils.JdbcUtils;
import example.model.Participant;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class ParticipantDBRepository implements ParticipantRepository {
    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public ParticipantDBRepository(Properties props) {
        logger.info("Initializing ParticipantDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void add(Participant entity) throws EntityRepoException {
        logger.traceEntry("add task {} ", entity);
        Connection connection = dbUtils.getConnection();
        String sql = "INSERT INTO Participants(name,age) VALUES (?,?)";
        try (var ps = connection.prepareStatement(sql)) {
            ps.setString(1, entity.getName());
            ps.setInt(2, entity.getAge());
            ps.executeUpdate();
            logger.traceExit("task {} added", entity);
        } catch (Exception e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
    }

    @Override
    public void remove(long id) throws EntityRepoException {
        logger.traceEntry("remove task {} ", id);
        Connection connection = dbUtils.getConnection();
        String sql = "DELETE FROM Participants WHERE id=?";
        try (var ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            ps.executeUpdate();
            logger.traceExit("task {} removed", id);
        } catch (Exception e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
    }

    @Override
    public List<Participant> getAll() throws EntityRepoException {
        logger.traceEntry("task getALL");
        Connection connection = dbUtils.getConnection();
        List<Participant> participants = new ArrayList<>();
        String sql = "SELECT * FROM Participants";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    Participant participant = extract(rs);
                    participants.add(participant);
                    logger.traceExit("task {} added", participant);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
        return participants;
    }

    @Override
    public Participant findById(long id) throws EntityRepoException {
        logger.traceEntry("task findById {}", id);
        Connection connection = dbUtils.getConnection();
        String sql = "SELECT * FROM Participants WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    Participant participant = extract(rs);
                    logger.traceExit("task {} found", participant);
                    return participant;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
        return null;
    }

    @Override
    public void update(long id, Participant entity) throws EntityRepoException {
        logger.traceEntry("update task {} ", entity);
        if (entity.getId() == null || entity.getId() < 0) {
            throw new EntityRepoException("Cannot update participant with null ID.");
        }
        Connection connection = dbUtils.getConnection();
        String sql = "UPDATE Participants SET name=?, age=? WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, entity.getName());
            ps.setInt(2, entity.getAge());
            ps.setLong(3, id);
            ps.executeUpdate();
            logger.traceExit("task {} updated", entity);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
    }

    private Participant extract(ResultSet resultSet) throws SQLException {
        Long id = resultSet.getLong("id");
        String name = resultSet.getString("name");
        int age = resultSet.getInt("age");
        Participant participant = new Participant(name, age);
        participant.setId(id);
        return participant;
    }

    @Override
    public Optional<Participant> getParticipantByData(Participant participant) {
        logger.traceEntry("task getParticipantByData {}", participant);
        Connection connection = dbUtils.getConnection();
        String sql = "SELECT * FROM Participants WHERE name=? AND age=?";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, participant.getName());
            preparedStatement.setInt(2, participant.getAge());
            try (ResultSet rs = preparedStatement.executeQuery()) {
                if (rs.next()) {
                    Participant part = extract(rs);
                    logger.traceExit("task {} found", participant);
                    return Optional.of(part);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        return Optional.empty();
    }

}

